File: packages/contracts/crypto/conformance/kotlin/src/main/kotlin/com/uankit53/crewroll/crypto/Binary.kt
package com.uankit53.crewroll.crypto

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest

enum class CryptoFailure {
    STRUCTURE,
    AUTHENTICATION,
    SEMANTIC_CONTEXT,
    CHECKSUM,
}

class CryptoReadException(
    val failure: CryptoFailure,
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause)

internal class BinaryReader(private val bytes: ByteArray) {
    private val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)

    val offset: Int
        get() = buffer.position()

    val remaining: Int
        get() = buffer.remaining()

    fun readBytes(count: Int): ByteArray {
        if (count < 0 || count > remaining) {
            throw CryptoReadException(
                CryptoFailure.STRUCTURE,
                "truncated binary input: need $count bytes, have $remaining",
            )
        }
        return ByteArray(count).also(buffer::get)
    }

    fun readU8(): Int = readBytes(1)[0].toInt() and 0xff

    fun readU16(): Int = ByteBuffer.wrap(readBytes(2))
        .order(ByteOrder.BIG_ENDIAN)
        .short
        .toInt() and 0xffff

    fun readU32(): Long = ByteBuffer.wrap(readBytes(4))
        .order(ByteOrder.BIG_ENDIAN)
        .int
        .toLong() and 0xffff_ffffL

    fun readU64(): ULong = readBytes(8).fold(0uL) { value, byte ->
        (value shl 8) or (byte.toInt() and 0xff).toULong()
    }

    fun assertEnd() {
        if (remaining != 0) {
            throw CryptoReadException(
                CryptoFailure.STRUCTURE,
                "trailing binary input: $remaining bytes",
            )
        }
    }
}

internal class BinaryWriter(capacity: Int) {
    private val buffer = ByteBuffer.allocate(capacity).order(ByteOrder.BIG_ENDIAN)

    fun writeAscii(value: String) {
        val bytes = value.toByteArray(Charsets.US_ASCII)
        if (bytes.size != value.length || bytes.any { (it.toInt() and 0xff) > 0x7f }) {
            throw CryptoReadException(
                CryptoFailure.STRUCTURE,
                "ASCII field contains non-ASCII",
            )
        }
        buffer.put(bytes)
    }

    fun writeU8(value: Int) {
        require(value in 0 .. 0xff)
        buffer.put(value.toByte())
    }

    fun writeU32(value: Long) {
        require(value in 0 .. 0xffff_ffffL)
        buffer.putInt(value.toInt())
    }

    fun toByteArray(): ByteArray = buffer.array().copyOf(buffer.position())
}

internal fun requireWidth(value: ByteArray, width: Int, label: String) {
    if (value.size != width) {
        throw CryptoReadException(
            CryptoFailure.STRUCTURE,
            "$label must be exactly $width bytes",
        )
    }
}

fun sha256(bytes: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(bytes)

internal fun bytesEqual(left: ByteArray, right: ByteArray): Boolean =
    MessageDigest.isEqual(left, right)

File: modules/crewroll-transfer/ios/IdentityKeys/Sources/CrewRollNativeKeys/MediaCrypto/Binary.swift
import Clibsodium
import Foundation

public typealias Bytes = [UInt8]

public enum CryptoReadError: Error, Equatable, LocalizedError {
    case structure(String)
    case authentication(String)
    case semanticContext(String)
    case checksum(String)

    public var errorDescription: String? {
        switch self {
        case let .structure(message),
             let .authentication(message),
             let .semanticContext(message),
             let .checksum(message):
            return message
        }
    }
}

struct ByteReader {
    private let bytes: Bytes
    private(set) var offset = 0

    init(_ bytes: Bytes) {
        self.bytes = bytes
    }

    var remaining: Int {
        bytes.count - offset
    }

    mutating func readBytes(_ count: Int) throws -> Bytes {
        guard count >= 0, count <= remaining else {
            throw CryptoReadError.structure(
                "truncated binary input: need \(count) bytes, have \(remaining)"
            )
        }
        let value = Bytes(bytes[offset ..< offset + count])
        offset += count
        return value
    }

    mutating func readUInt8() throws -> UInt8 {
        try readBytes(1)[0]
    }

    mutating func readUInt16() throws -> UInt16 {
        let value = try readBytes(2)
        return UInt16(value[0]) << 8 | UInt16(value[1])
    }

    mutating func readUInt32() throws -> UInt32 {
        let value = try readBytes(4)
        return value.reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    }

    mutating func readUInt64() throws -> UInt64 {
        let value = try readBytes(8)
        return value.reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
    }

    mutating func assertEnd() throws {
        guard remaining == 0 else {
            throw CryptoReadError.structure(
                "trailing binary input: \(remaining) bytes"
            )
        }
    }
}

struct ByteWriter {
    private(set) var bytes: Bytes = []
    mutating func erase() { sodium_memzero(&bytes, bytes.count) }

    mutating func write(_ value: UInt8) {
        bytes.append(value)
    }

    mutating func write(_ value: UInt32) {
        bytes.append(UInt8(truncatingIfNeeded: value >> 24))
        bytes.append(UInt8(truncatingIfNeeded: value >> 16))
        bytes.append(UInt8(truncatingIfNeeded: value >> 8))
        bytes.append(UInt8(truncatingIfNeeded: value))
    }

    mutating func write(_ value: Bytes) {
        bytes.append(contentsOf: value)
    }

    mutating func writeASCII(_ value: String) throws {
        let encoded = Bytes(value.utf8)
        guard encoded.allSatisfy({ $0 <= 0x7f }) else {
            throw CryptoReadError.structure("ASCII field contains non-ASCII")
        }
        write(encoded)
    }
}

func requireWidth(_ value: Bytes, _ width: Int, _ label: String) throws {
    guard value.count == width else {
        throw CryptoReadError.structure(
            "\(label) must be exactly \(width) bytes"
        )
    }
}

func bytesEqual(_ left: Bytes, _ right: Bytes) -> Bool {
    guard left.count == right.count else {
        return false
    }
    var difference: UInt8 = 0
    for index in left.indices {
        difference |= left[index] ^ right[index]
    }
    return difference == 0
}

public func sha256(_ bytes: Bytes) throws -> Bytes {
    var digest = Bytes(repeating: 0, count: Int(crypto_hash_sha256_bytes()))
    guard crypto_hash_sha256(&digest, bytes, UInt64(bytes.count)) == 0 else {
        throw CryptoReadError.authentication("libsodium SHA-256 failed")
    }
    return digest
}

public enum CrewRollSodium {
    public static func initialize() throws {
        guard sodium_init() >= 0 else {
            throw CryptoReadError.authentication("sodium_init failed")
        }
    }

    public static var version: String {
        String(cString: sodium_version_string())
    }
}
